### DISCUSSION TITLE
Push Notifications V2

### DISCUSSION THREAD

**Discussion by VNDRN** [GetStream/stream-chat-react-native#2343] [Q&A] [answered]
I’m trying to implement push notifications by following [the guide in the docs](https://getstream.io/chat/docs/sdk/reactnative/guides/push-notifications-v2/). I’ve encountered some issues/questions whilst doing. 

In the section Upload Firebase Credentials under the dashboard section, this seems to be outdated? You can now access it via **dashboard > Chat Messaging > Push Notifications** instead of the description and images in the guide.

In the section **Registering a device with Stream** , it is unclear to me what the difference is between `client.setLocalDevice` and `client.addDevice`. I believe `setLocalDevice` was added for V2, but could someone explain the difference to me? 

Thanks!

**Comment by vishalnarkhede** [comment, marked as answer]
Prior to `client.setLocalDevice`, you would need to do two separate api calls:

```ts
await client.connectUser();
await client.addDevice();
```

But later we decided to make this more efficient, and merge these two api calls:

```ts
client.setLocalDevice(); // there is no api call here, we just store the device info in memory
await client.connectUser(); // pass the device info to backend
```

**Reply by VNDRN** [reply]
Ok, I get that, but why call client.addDevice after the user disconnects? It happens in the unsubscribeTokenRefreshListenerRef, which is called after client.disconnect?

And I could still use addDevice to add the device after the connection is already established?

**Reply by santhoshvai** [reply]
Hey @VNDRN 

it is not done after user disconnecting.. but rather after token refreshing (this is done by firebase)

```ts
messaging().onTokenRefresh(async newToken => {
        await Promise.all([
          removeOldToken(),
          client.addDevice(newToken, push_provider, USER_ID, push_provider_name),
          AsyncStorage.setItem('@current_push_token', newToken),
        ]);
      });
    };
```

This method will return a subscription removal method, and that is stored inside `unsubscribeTokenRefreshListenerRef`

What we do in this method is to remove the stale token from the server and send the new token 

I hope its clear now
